devito/types/tensor.py

```diff
@@ -0,0 +1,601 @@
+        func = tens_func(self, self)
+        return func(name='%s%s' % (self.name, func.name),
+                    grid=self.grid, space_order=self.space_order,
+                    components=comps, time_order=to, symmetric=self.is_symmetric,
+                    diagonal=self.is_diagonal)
+
+    @classmethod
+    def __dtype_setup__(cls, **kwargs):
+        return Function.__dtype_setup__(**kwargs)
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        return Function.__indices_setup__(grid=kwargs.get('grid'),
+                                          dimensions=kwargs.get('dimensions'))
+
+    @property
+    def is_diagonal(self):
+        return self._is_diagonal
+
+    @property
+    def is_symmetric(self):
+        return self._is_symmetric
+
+    @property
+    def indices(self):
+        return self._indices
+
+    @property
+    def staggered(self):
+        return self._staggered
+
+    @property
+    def space_dimensions(self):
+        return self.indices
+
+    @property
+    def grid(self):
+        return self._grid
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def space_order(self):
+        return self._space_order
+
+    @property
+    def evaluate(self):
+        return self.applyfunc(lambda x: getattr(x, 'evaluate', x))
+
+    # Custom repr and str
+    def __str__(self):
+        name = "SymmetricTensor" if self._is_symmetric else "Tensor"
+        if self._is_diagonal:
+            name = "DiagonalTensor"
+        st = ''.join([' %-2s,' % c for c in self.values()])
+        return "%s(%s)" % (name, st)
+
+    __repr__ = __str__
+
+    def _sympy_(self):
+        return self
+
+    @classmethod
+    def _sympify(cls, arg):
+        return arg
+
+    def _entry(self, i, j, **kwargs):
+        return self.__getitem__(i, j)
+
+    def __getitem__(self, *args):
+        if len(args) == 1:
+            return super(TensorFunction, self).__getitem__(*args)
+        i, j = args
+        if self.is_diagonal:
+            if i == j:
+                return super(TensorFunction, self).__getitem__(i, j)
+            return 0.0
+        if self.is_symmetric:
+            if j < i:
+                return super(TensorFunction, self).__getitem__(j, i)
+            else:
+                return super(TensorFunction, self).__getitem__(i, j)
+        return super(TensorFunction, self).__getitem__(i, j)
+
+    def _eval_transpose(self):
+        if self.is_symmetric or self.is_diagonal:
+            return self
+        mat = [[self[j, i] for j in range(self.cols)] for i in range(self.rows)]
+        func = tens_func(self, self)
+        name = '%s_T' % self.name
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=mat, time_order=to, symmetric=self.is_symmetric,
+                    diagonal=self.is_diagonal)
+
+    def values(self, symmetric=False):
```

Comment:
Why is there a `symmetric` optional argument in addition to the `is_symmetric` property? To get the diagonal?
Maybe I missed it, but I can't see a `.values(symmetric=True)` in the current code base.

Reply:
`.values(symmetric=True)` extra values "as if it was symmetric" but it may not be symmetric, while `is_symmetric` defines whether or not the tensor is symmetric

Reply:
Oh, I see. But then I think `symmetrize` would be a better name. Also, why (if I'm reading it correctly) is there only the option to symmetrize via the UD?